src/server/pkg/storage/fileset/writer.go

```diff
@@ -44,19 +55,47 @@ func (w *Writer) WriteHeader(hdr *index.Header) error {
 		Hdr: hdr.Hdr,
 		Idx: &index.Index{DataOp: &index.DataOp{}},
 	}
-	w.cw.StartRange(w.callback(w.hdr))
+	w.cw.Annotate(&chunk.Annotation{
+		NextDataRef: &chunk.DataRef{},
+		Meta: &meta{
+			hdr: w.hdr,
+		},
+	})
 	if err := w.tw.WriteHeader(w.hdr.Hdr); err != nil {
 		return err
 	}
 	// Setup first tag for header.
-	w.hdr.Idx.DataOp.Tags = []*index.Tag{&index.Tag{Id: headerTag, SizeBytes: w.cw.RangeSize()}}
+	w.hdr.Idx.DataOp.Tags = []*index.Tag{&index.Tag{Id: headerTag, SizeBytes: w.cw.AnnotationSize()}}
 	return nil
 }
 
-func (w *Writer) callback(hdr *index.Header) func([]*chunk.DataRef) error {
-	return func(dataRefs []*chunk.DataRef) error {
-		hdr.Idx.DataOp.DataRefs = dataRefs
-		return w.iw.WriteHeader(hdr)
+func (w *Writer) callback() chunk.WriterFunc {
+	return func(_ *chunk.DataRef, annotations []*chunk.Annotation) error {
+		hdr := annotations[0].Meta.(*meta).hdr
```

Comment:
Why is it necessary to do a cast here? can `Annotations.Meta` not be of the right type? Is there any worry about the wrong type getting in here and this cast failing?

Reply:
Imo this would be better as a concrete type, `interface{}` says nothing about the type being passed so I generally only feel like it's reasonable when you can't pin down a specific, like with a generic sort function. When it's just being used to allow for two different types I think that's a sign that those types should be a single variadic type, or possible support a common, non-empty, interface.

Reply:
The way I am looking at it is that an annotation can have some additional arbitrary information associated. There really is not anything to say about what is there from the chunk package's perspective because it does not make use of it. It is only relevant to the client package that is using the chunk package.